sktime/transformations/base.py

```diff
@@ -48,93 +104,567 @@ class BaseTransformer(BaseEstimator):
+        ]
+
+        # convert X and y to a supported internal type
+        #  if X/y type is already supported, no conversion takes place
+        X_inner = convert_to(
+            X,
+            to_type=X_inner_mtype,
+            as_scitype=X_scitype,
+        )
+        y_inner = convert_to(
+            y,
+            to_type=y_inner_mtype,
+            as_scitype=y_input_scitype,
+        )
+
+        # todo: uncomment this once Z is completely gone
+        # self._fit(X=X_inner, y=y_inner)
+        # less robust workaround until then
+        self._fit(X_inner, y_inner)
+
         self._is_fitted = True
         return self
 
-    def transform(self, Z, X=None):
-        """Transform data. Returns a transformed version of Z."""
-        raise NotImplementedError("abstract method")
+    def transform(self, X, y=None, Z=None):
+        """Transform X and return a transformed version.
+
+        State required:
+            Requires state to be "fitted".
+
+        Accesses in self:
+            Fitted model attributes ending in "_".
+            self._is_fitted
+
+        Parameters
+        ----------
+        X : Series or Panel, any supported mtype
+            Data to be transformed, of python type as follows:
+                Series: pd.Series, pd.DataFrame, or np.ndarray (1D or 2D)
+                Panel: pd.DataFrame with 2-level MultiIndex, list of pd.DataFrame,
+                    nested pd.DataFrame, or pd.DataFrame in long/wide format
+                subject to sktime mtype format specifications, for further details see
+                    examples/AA_datatypes_and_datasets.ipynb
+        y : Series or Panel, optional (default=None)
+            Additional data, e.g., labels for transformation
+        Z : possible alias for X; should not be passed when X is passed
+            alias Z will be deprecated in version 0.10.0
+
+        Returns
+        -------
+        transformed version of X
+        type depends on type of X and scitype:transform-output tag:
+            |          | `transform`  |                        |
+            |   `X`    |  `-output`   |     type of return     |
+            |----------|--------------|------------------------|
+            | `Series` | `Primitives` | `pd.DataFrame` (1-row) |
+            | `Panel`  | `Primitives` | `pd.DataFrame`         |
+            | `Series` | `Series`     | `Series`               |
+            | `Panel`  | `Series`     | `Panel`                |
+            | `Series` | `Panel`      | `Panel`                |
+        instances in return correspond to instances in `X`
+        combinations not in the table are currently not supported
+        """
+        X = _handle_alias(X, Z)
+
+        # check whether is fitted, unless fit-in-transform is true
+        if self.get_tag("fit-in-transform"):
+            self.fit(X=X, y=y, Z=Z)
+        else:
+            self.check_is_fitted()
+
+        # input checks and minor coercions on X, y
```

Comment:
Up to you but if these comments are here because this method is hard to follow potentially move each of the logics under these comment blocks to their own method and call them.

Reply:
well, I'm wondering how you would split this up into functions so it improves readability. If you move things out, it might make things less readable? Suggestions appreciated.

Reply:
Well im not sure myself either - if you think it's more obvious like this then go with this.